### PR TITLE
Adds 'stub_email_alert_api_has_subscriptions' test method

### PR DESCRIPTION
**Pull request by ChrisBAshton**
Trello card: https://trello.com/c/VlnvU2Dl
Depended on by: https://github.com/alphagov/email-alert-frontend/pull/474 (please read that for more context)

## What
- Adds a `stub_email_alert_api_has_subscriptions` method, which takes an array of subscriptions to stub, so that we can test [like so](https://github.com/alphagov/email-alert-frontend/pull/474/files#diff-20c575b223c5b129c1750d0f400abda8R64).
- Ensures that `stub_email_alert_api_has_subscription` stubs the '/latest' endpoint (this is needed if we're only stubbing one subscription response).
- Refactors the shared code between these methods into a new `fill_gaps_with_default_values` method.
- Writes tests for `stub_email_alert_api_has_subscriptions`.

## Why
We added a `get_latest_matching_subscription` method in https://github.com/alphagov/gds-api-adapters/pull/928 (v59.3.0), which takes a subscription ID and returns the latest matching subscription response (e.g. a user changing frequency from Daily to Weekly, if we query '/latest' on the Daily subscription, it will return the Weekly one as that is the most recent).

Consequently, the return value of '/latest' depends upon whether or not a newer subscription has been stubbed. It was very difficult / hacky / impossible to test the '/latest' method behaviour with the `get_latest_matching_subscription` method alone, so I've added a `stub_email_alert_api_has_subscriptions` which takes an array of subscriptions and stubs them in the order they are added.

I.e. if you pass `stub_email_alert_api_has_subscriptions([subscription1, subscription2])` and then query '/latest' on `subscription1`, the response for `subscription2` should be returned.

## Note
It was quite difficult to write 'clean' code here because the `stub_email_alert_api_has_subscription` method takes three parameters (`id`, `frequency`, `'hash of options'`), which means you cannot pass `{ id: 1, frequency: 'weekly', ended: false }` etc directly but have to instead remove the options that are not explicitly allowed in that third parameter.

Consequently, I was unable to simply call `stub_email_alert_api_has_subscription` from within `stub_email_alert_api_has_subscriptions` without hacking the options parameter, and it was cleaner to simply implement the `stub_request` method in place. I also have had to return `[id, parameters]`, rather than simply `parameters` (containing `:id`) as again, this otherwise leads to a `ArgumentError: unknown keyword: id`.

In the long term, it would be great to update `stub_email_alert_api_has_subscription` to take a single hash rather than 3 params, but this will involve a change to the email-alert-frontend tests which rely upon the method. Doing this should delete several lines of code and improve code readability and consistency.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,8 @@
+# 59.3.1
```
Comment (leenagupte) [inline CHANGELOG.md]: Updates to the version number should happen in a separate PR. For now, the changes should be listed as `# Unreleased`

Comment (ChrisBAshton) [reply CHANGELOG.md]: Oops - good to know! Will do that.

File: lib/gds_api/test_helpers/email_alert_api.rb [hunk 2]
```diff
@@ -45,26 +45,25 @@ def stub_email_alert_api_does_not_have_subscriber_subscriptions(id)
...
 
-      def stub_email_alert_api_has_subscription(
-        id,
-        frequency,
-        title: "Some title",
-        subscriber_id: 1,
-        subscriber_list_id: 1000,
-        ended: false
-      )
+      def stub_email_alert_api_has_subscription(id, frequency, optional_overrides = {})
```
Comment (leenagupte) [inline lib/gds_api/test_helpers/email_alert_api.rb]: @ChrisBAshton The method already allows you to override the values for title, subscriber_id, subscriber_list_id, and ended. Leaving them as named parameters makes it easier to find out what these values are.
It also means that you don't need a separate method to add these values.

File: lib/gds_api/test_helpers/email_alert_api.rb [hunk 3]
```diff
@@ -227,6 +226,7 @@ def stub_email_alert_api_does_not_have_subscriber_list_by_slug(slug:)
       alias_method :email_alert_api_has_subscriber_subscriptions, :stub_email_alert_api_has_subscriber_subscriptions
       alias_method :email_alert_api_does_not_have_subscriber_subscriptions, :stub_email_alert_api_does_not_have_subscriber_subscriptions
       alias_method :email_alert_api_has_subscription, :stub_email_alert_api_has_subscription
+      alias_method :email_alert_api_has_subscriptions, :stub_email_alert_api_has_subscriptions
```
Comment (leenagupte) [inline lib/gds_api/test_helpers/email_alert_api.rb]: Was this missed in the previous PR?

File: lib/gds_api/test_helpers/email_alert_api.rb [hunk 4]
```diff
@@ -270,7 +270,7 @@ def get_subscription_response(
             "id" => id,
             "frequency" => frequency,
             "source" => "user_signed_up",
-            "ended_at" => ended ? Time.now.rfc3339 : nil,
+            "ended_at" => ended ? Time.now.to_datetime.rfc3339 : nil,
```
Comment (leenagupte) [inline lib/gds_api/test_helpers/email_alert_api.rb]: Why has this changed? Can you move it to a separate commit and add an explanation?

File: lib/gds_api/test_helpers/email_alert_api.rb [hunk 5]
```diff
@@ -45,26 +45,25 @@ def stub_email_alert_api_does_not_have_subscriber_subscriptions(id)
...
-              id,
-              frequency: frequency,
-              title: title,
-              subscriber_id: subscriber_id,
-              subscriber_list_id: subscriber_list_id,
-              ended: ended,
-            ).to_json,
-          )
+          .to_return(status: 200, body: response)
+        stub_request(:get, "#{EMAIL_ALERT_API_ENDPOINT}/subscriptions/#{id}/latest")
```
Comment (leenagupte) [inline lib/gds_api/test_helpers/email_alert_api.rb]: Should this be in a separate method? i.e. `stub_email_alert_api_has_latest_subscription`?

File: lib/gds_api/test_helpers/email_alert_api.rb (hunk 3 above)
Comment (ChrisBAshton) [reply lib/gds_api/test_helpers/email_alert_api.rb]: No, the previous PR added a new API method that can be used by email-alert-frontend.

This PR adds a new test helper, which needs exposing here.

File: lib/gds_api/test_helpers/email_alert_api.rb (hunk 2 above)
Comment (ChrisBAshton) [reply lib/gds_api/test_helpers/email_alert_api.rb]: The list of keys was in two places (the `stub_email_alert_api_has_subscription` method parameters, and the `get_subscription_response` call below it), and in creating my `stub_email_alert_api_has_subscriptions` method I risked duplicating it a third time and the default values a second time (in my `get_subscription_response` call).

Felt like a good idea to extract that out so that we only have one place to update if we want to change a default value or add a new override. If we'd prefer the duplication I can do that.

File: lib/gds_api/test_helpers/email_alert_api.rb (hunk 5 above)
Comment (ChrisBAshton) [reply lib/gds_api/test_helpers/email_alert_api.rb]: That crossed my mind, but _every_ subscription should have a valid response if you call `/latest`. So it would mean otherwise writing this in our tests:

```ruby
stub_email_alert_api_has_subscription(1, 'weekly', :ended)
stub_email_alert_api_has_latest_subscription(1, 'weekly', :ended)
stub_email_alert_api_has_subscription(2, 'weekly')
stub_email_alert_api_has_latest_subscription(2, 'weekly')
```

I think this makes things less readable - it can be misinterpreted as "this subscription should be the latest one", which is not what we want. What we want, if anything, is "this subscription should be queryable against the '/latest' endpoint". But if you've stubbed the email API for a subscription, that should just be done for you IMO.

What do you think?

Comment (leenagupte) [reply lib/gds_api/test_helpers/email_alert_api.rb]: Can you add this explanation to the commit note?

File: lib/gds_api/test_helpers/email_alert_api.rb (hunk 2 above)
Comment (leenagupte) [reply lib/gds_api/test_helpers/email_alert_api.rb]: At this point (the first commit) you only have one method (`stub_email_alert_api_has_subscription`) so extracting out the default values it's necessary.

Comment (ChrisBAshton) [reply lib/gds_api/test_helpers/email_alert_api.rb]: Ah, agreed. Should I split it out so that it's repeated initially, then refactored into its own method in a subsequent commit?

File: lib/gds_api/test_helpers/email_alert_api.rb [hunk 6]
```diff
@@ -57,6 +57,21 @@ def stub_email_alert_api_has_subscription(id, frequency, optional_overrides = {}
           .to_return(status: 200, body: response)
       end
 
+      # Stubs the API responses as if each subscription happened in the order they are passed
+      def stub_email_alert_api_has_subscriptions(*subscriptions)
+        subscriptions.map! do |id, frequency, optional_overrides|
```
Comment (leenagupte) [inline lib/gds_api/test_helpers/email_alert_api.rb]: Rather that specifying the parameters in the block definition, you can just use 
`subscriptions.map! do |subscription|` and use `subscription[:id]`  to populate the id.

File: lib/gds_api/test_helpers/email_alert_api.rb [hunk 7]
```diff
@@ -57,6 +57,21 @@ def stub_email_alert_api_has_subscription(id, frequency, optional_overrides = {}
           .to_return(status: 200, body: response)
       end
 
+      # Stubs the API responses as if each subscription happened in the order they are passed
+      def stub_email_alert_api_has_subscriptions(*subscriptions)
```
Comment (leenagupte) [inline lib/gds_api/test_helpers/email_alert_api.rb]: What is the benefit of using the a splat here?

File: lib/gds_api/test_helpers/email_alert_api.rb (hunk 6 above)
Comment (ChrisBAshton) [reply lib/gds_api/test_helpers/email_alert_api.rb]: I don't think you can? `subscription` is a tuple of two strings and a hash:

```ruby
[
  'abc123',
  'daily',
  {
    ended: false,
    subscription_reason: 'etc',
  }
]
```

Comment (leenagupte) [reply lib/gds_api/test_helpers/email_alert_api.rb]: How would you get the `id` for `subscriptions[0]`? That's what you'd use in the loop.

File: lib/gds_api/test_helpers/email_alert_api.rb (hunk 3 above)
Comment (leenagupte) [reply lib/gds_api/test_helpers/email_alert_api.rb]: I think this is in the wrong commit then, as `stub_email_alert_api_has_subscriptions` doesn't yet.

File: test/email_alert_api_test.rb [hunk 8]
```diff
@@ -79,6 +79,48 @@
...
+    describe "user unsubscribed then resubscribed at the same frequency" do
+      before do
+        stub_email_alert_api_has_subscriptions(
+          [1, "weekly", ended: true],
+          [2, "weekly"]
+        )
+      end
+
+      it "returns the latest subscription attributes on /latest" do
+        subscription_attrs = api_client.get_latest_matching_subscription(1) # notice the 1
```
Comment (leenagupte) [inline test/email_alert_api_test.rb]: Rather use a comment, could the ids be changed to something more obvious? i.e. a larger number?

File: test/email_alert_api_test.rb [hunk 9]
```diff
@@ -79,6 +79,48 @@
...
+          .fetch("subscription")
+
+        assert_equal(1, subscription_attrs.fetch("id"))
+      end
+    end
+
+    describe "user unsubscribed then resubscribed at the same frequency" do
+      before do
+        stub_email_alert_api_has_subscriptions(
+          [1, "weekly", ended: true],
```
Comment (leenagupte) [inline test/email_alert_api_test.rb]: What happens if the user has more than one active subscription?

File: test/email_alert_api_test.rb [hunk 10]
```diff
@@ -79,6 +79,48 @@
 
         assert_equal("weekly", subscription_attrs.fetch("frequency"))
       end
+
+      it "returns the subscription attributes on /latest" do
```
Comment (leenagupte) [inline test/email_alert_api_test.rb]: This test looks like a duplicate of line 100 `it "returns the latest subscription attributes on /latest"` are they both needed?

File: test/email_alert_api_test.rb (hunk 9 above)
Comment (leenagupte) [reply test/email_alert_api_test.rb]: Also what happens if a user's subscriptions look like: 

```
[1, "weekly"],
 [2, "daily", ended: true]
```

File: lib/gds_api/test_helpers/email_alert_api.rb (hunk 7 above)
Comment (leenagupte) [inline lib/gds_api/test_helpers/email_alert_api.rb]: I think you can justify expanding the comment on the method to clarify what a `subscriptions` object looks like, and how the method should be called. 
There are some examples in the publishing-api test helpers: https://github.com/alphagov/gds-api-adapters/blob/master/lib/gds_api/test_helpers/publishing_api_v2.rb

File: lib/gds_api/test_helpers/email_alert_api.rb (hunk 2 above)
Comment (leenagupte) [reply lib/gds_api/test_helpers/email_alert_api.rb]: I think it's ok to keep the duplication. It's likely that the two test helpers will be used independently of each other, so I think it'll be more helpful to the person using them to be able to clearly see the defaults.

File: lib/gds_api/test_helpers/email_alert_api.rb (hunk 6 above)
Comment (ChrisBAshton) [reply lib/gds_api/test_helpers/email_alert_api.rb]: If I do `subscriptions.map! do |subscription|` then inside the loop it will either be a):

```ruby
id, frequency, optional_overrides = subscription
```

or b):

```ruby
id = subscription[0]
frequency = subscription[1]
optional_overrides = subscription[2]
```

Which would be best?

File: lib/gds_api/test_helpers/email_alert_api.rb (hunk 7 above)
Comment (ChrisBAshton) [reply lib/gds_api/test_helpers/email_alert_api.rb]: Good point! None :)

Comment (ChrisBAshton) [reply lib/gds_api/test_helpers/email_alert_api.rb]: Because we pass subscriptions as a 0..* list, not an array

```ruby
stub_email_alert_api_has_subscriptions(
  [1, 'weekly', :ended]
  [2, 'daily']
)
```

Happy to change the API to take an array instead, i.e.

```ruby
stub_email_alert_api_has_subscriptions([
  [1, 'weekly', :ended]
  [2, 'daily']
])
```

File: lib/gds_api/test_helpers/email_alert_api.rb (hunk 6 above)
Comment (leenagupte) [reply lib/gds_api/test_helpers/email_alert_api.rb]: Could `subscriptions` be changed to be an array of hashes?
e.g.
```
[
  {
    id: 'abc123', 
    frequency: 'daily',
    ended: false,
    subscription_reason: 'etc',
  }
]
```
It might make it easier to call from a test.

File: test/email_alert_api_test.rb (hunk 9 above)
Comment (leenagupte) [reply test/email_alert_api_test.rb]: That's not true. At the moment you can create a daily and weekly etc, subscription to the same list.

Comment (ChrisBAshton) [reply test/email_alert_api_test.rb]: I thought one automatically unsubscribed the other?

Comment (leenagupte) [reply test/email_alert_api_test.rb]: I thought @brucebolt  created two yesterday. It's worth checking, and adding a test if required and a commit note if not.

Comment (ChrisBAshton) [reply test/email_alert_api_test.rb]: Bruce & I just checked this; I signed up for Daily and then Immediately on integration (both from the `/email/subscriptions/new?topic_id=find-eu-exit-guidance-for-your-business-appear-in-find-eu-exit-guidance-business-finder` endpoint).

In the database, the Daily one was `ended` with reason `frequency_changed`, and the Immediately one has a 'source' of `frequency_changed`, so it doesn't look like it's possible to have multiple active subscriptions.

But we did look for a test and couldn't find one that captures this, so I'll add one in a PR 👍

File: lib/gds_api/test_helpers/email_alert_api.rb (hunk 4 above)
Comment (ChrisBAshton) [reply lib/gds_api/test_helpers/email_alert_api.rb]: Done.

File: lib/gds_api/test_helpers/email_alert_api.rb (hunk 3 above)
Comment (ChrisBAshton) [reply lib/gds_api/test_helpers/email_alert_api.rb]: Whoops!

File: test/email_alert_api_test.rb (hunk 9 above)
Comment (ChrisBAshton) [reply test/email_alert_api_test.rb]: Oh actually it is being tested here: https://github.com/alphagov/email-alert-api/blob/8c68924f632760c22154757e54b903841fb7a86d/spec/integration/create_subscription_spec.rb#L35-L58

File: lib/gds_api/test_helpers/email_alert_api.rb [hunk 11]
```diff
@@ -68,6 +68,54 @@ def stub_email_alert_api_has_subscription(
...
+      #      ended: true,
+      #    },
+      #    {
+      #      id: 'id-of-my-subscriber-list',
+      #      frequency: 'daily',
+      #    },
+      #  ])
+      #
+      # @param subscriptions [Array]
+      def stub_email_alert_api_has_subscriptions(subscriptions)
```
Comment (leenagupte) [inline lib/gds_api/test_helpers/email_alert_api.rb]: As the last subscription in the list should be the "latest" I wonder whether there's any point in passing `ended` into the hash, and instead automatically set it to true on all but the last subscription in the list?

File: lib/gds_api/test_helpers/email_alert_api.rb [hunk 12]
```diff
@@ -68,6 +68,54 @@ def stub_email_alert_api_has_subscription(
...
+        subscriptions.map! { |subscription| apply_subscription_defaults(subscription) }
+        subscriptions.each do |id, params|
```
Comment (leenagupte) [inline lib/gds_api/test_helpers/email_alert_api.rb]: Are two loops needed here?

File: lib/gds_api/test_helpers/email_alert_api.rb (hunk 11 above)
Comment (ChrisBAshton) [reply lib/gds_api/test_helpers/email_alert_api.rb]: Still worth passing it to make the tests clear, I think - but can definitely override all but the last subscription to be `ended: true`. 👍

File: lib/gds_api/test_helpers/email_alert_api.rb (hunk 12 above)
Comment (ChrisBAshton) [reply lib/gds_api/test_helpers/email_alert_api.rb]: Yes, as we pass the modified `subscriptions` as a parameter on the next line. We've got a nice API for calling the method, and we need to transform the data into something we can pass around.

Having two loops as we do here is a complexity of `2n`; the alternative is a loop within a loop (`n²`).

File: lib/gds_api/test_helpers/email_alert_api.rb (hunk 11 above)
Comment (ChrisBAshton) [reply lib/gds_api/test_helpers/email_alert_api.rb]: I now set `ended: true` by default in `apply_subscription_defaults`.

File: lib/gds_api/test_helpers/email_alert_api.rb (hunk 12 above)
Comment (brucebolt) [reply lib/gds_api/test_helpers/email_alert_api.rb]: Would it be possible to have something like this to avoid two loops?

```
subscriptions.each do |subscription|
  subscription = apply_subscription_defaults(subscription)
  ...
```

File: lib/gds_api/test_helpers/email_alert_api.rb [hunk 13]
```diff
@@ -68,6 +68,54 @@ def stub_email_alert_api_has_subscription(
...
+      end
+
+      def apply_subscription_defaults(subscription)
+        parameters = {
+          title: "Some title",
+          subscriber_id: 1,
+          subscriber_list_id: 1000,
+          ended: true,
+        }.merge(subscription)
+        id = parameters.delete(:id) # Otherwise `ArgumentError: unknown keyword: id`
```
Comment (brucebolt) [inline lib/gds_api/test_helpers/email_alert_api.rb]: I'm not clear on why the `id` value has to be deleted from `parameters`.  Where does the `ArgumentError` occur if this isn't deleted?

File: lib/gds_api/test_helpers/email_alert_api.rb [hunk 14]
```diff
@@ -68,6 +68,54 @@ def stub_email_alert_api_has_subscription(
...
+        [id, parameters]
+      end
+
+      def get_latest_matching(params, subscriptions)
+        matching = subscriptions.select do |_current_id, current_params|
+          params[:subscriber_id] == current_params[:subscriber_id] &&
+            params[:subscriber_list_id] == current_params[:subscriber_list_id]
+        end
+        matching.last
```
Comment (brucebolt) [inline lib/gds_api/test_helpers/email_alert_api.rb]: Do you need to specify the ordering of results for the `select` query so `.last` is always the final result?  I can't see a `default_scope` for the Subscription model, so the return of results may not always be in a consistent order.

File: lib/gds_api/test_helpers/email_alert_api.rb (hunk 13 above)
Comment (leenagupte) [inline lib/gds_api/test_helpers/email_alert_api.rb]: Is the description being merged in from the subscription? Is it better to be explicit about what can be added to the parameters hash? i.e. only adding `frequency` and `ended` if `ended` exists?

File: lib/gds_api/test_helpers/email_alert_api.rb (hunk 12 above)
Comment (ChrisBAshton) [reply lib/gds_api/test_helpers/email_alert_api.rb]: Not reeeeeally, because we'd be using `subscriptions` before all of the subscriptions have been transformed, like:

```ruby
subscriptions.each do |subscription|
  subscription = apply_subscription_defaults(subscription)
  foo(subscriptions) # oops - we now need to apply defaults to all of these somehow
```

Which is why I've done it outside the loop:

```ruby
subscriptions = transform_the_subs(subscriptions)
subscriptions.each do |subscription|
  subscription = apply_subscription_defaults(subscription)
  foo(subscriptions) # now we can call this safely :)
```

If it's code _efficiency_ we're worried about, I think it's better as-is, otherwise within `foo(subscriptions)` we have to call `apply_subscription_defaults` on each subscription (a loop within a loop).

If it's just readability we're worried about, I can look at calling `apply_subscription_defaults` in the `get_latest_matching` method...?

File: lib/gds_api/test_helpers/email_alert_api.rb (hunk 13 above)
Comment (ChrisBAshton) [reply lib/gds_api/test_helpers/email_alert_api.rb]: The id hack is in order to support the nice API you suggested here: https://github.com/alphagov/gds-api-adapters/pull/932#discussion_r286040228

Unfortunately I can't pass the parameters 'blindly' to `get_subscription_response` as it complains that it doesn't recognise `:id`.

The allowed parameters are in the defaults (`title`, `subscriber_id`, `subscriber_list_id`, `ended`), apart from `id` (which will be stripped out) and `frequency`. If `frequency` is omitted it will error, and anything extra that's added accidentally will cause the `ArgumentError`. Could have had a default for 'frequency' but think it's better for it to be explicitly provided every time. Hope that makes sense!

File: lib/gds_api/test_helpers/email_alert_api.rb (hunk 14 above)
Comment (ChrisBAshton) [reply lib/gds_api/test_helpers/email_alert_api.rb]: > The return of results may not always be in time order.

What do you mean by this? Does Ruby not guarantee the order of arrays being traversed? I know in JavaScript you can't rely on the order of JSON properties being traversed, but don't expect that to be a problem here...?

The `stub_email_alert_api_has_subscriptions` works on the basis that the subscriptions are applied in the order they are passed, i.e. you should not be putting the most recent subscription halfway through the arguments.

File: lib/gds_api/test_helpers/email_alert_api.rb (hunk 13 above)
Comment (ChrisBAshton) [reply lib/gds_api/test_helpers/email_alert_api.rb]: Fair point - I'll make the comment more descriptive 👍

File: lib/gds_api/test_helpers/email_alert_api.rb (hunk 14 above)
Comment (brucebolt) [reply lib/gds_api/test_helpers/email_alert_api.rb]: I see, so the input to `stub_email_alert_api_has_subscriptions` will always be in the correct order, then `.last` is fine here.

File: lib/gds_api/test_helpers/email_alert_api.rb (hunk 13 above)
Comment (ChrisBAshton) [reply lib/gds_api/test_helpers/email_alert_api.rb]: Have tweaked the comment to hopefully make this clearer.

File: lib/gds_api/test_helpers/email_alert_api.rb (hunk 12 above)
Comment (brucebolt) [reply lib/gds_api/test_helpers/email_alert_api.rb]: Ah, yes, I see now.  I cannot see any way the two loops could be condensed into one without any further code complexity elsewhere.
